lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -425,6 +438,202 @@ prop_availableUTxO makeProperty =
+-- verifies that all entries match the condition that was provided.
+--
+prop_changeUTxO :: Property
+prop_changeUTxO =
+    forAllShrink (listOf genTxOutputs) (shrinkList shrinkTxOutputs) inner
+  where
+    inner :: [TxOutputs] -> Property
+    inner txOutputs =
+        checkCoverage $
+        cover 50 (not (UTxO.null utxoEven) && not (UTxO.null utxoOdd))
+            "UTxO sets not null" $
+        conjoin
+            [ -- All addresses in the even-parity UTxO set have even parity:
+              F.all ((== Even) . txOutParity) (unUTxO utxoEven)
+              -- All addresses in the odd-parity UTxO set have odd parity:
+            , F.all ((== Odd) . txOutParity) (unUTxO utxoOdd)
+              -- The even-parity and odd-parity UTxO sets are disjoint:
+            , Map.null $ Map.intersection (unUTxO utxoEven) (unUTxO utxoOdd)
+              -- The even-parity and odd-parity UTxO sets are complete:
+            , Map.union (unUTxO utxoEven) (unUTxO utxoOdd) == unUTxO utxoAll
+              -- No outputs are omitted when we select everything:
+            , UTxO.size utxoAll == F.sum (F.length . unTxOutputs <$> txOutputs)
+            ]
+      where
+        -- Computes the parity of an output based on its address parity.
+        txOutParity :: TxOut -> Parity
+        txOutParity = addressParity . view #address
+
+        -- The UTxO set that contains all available output addresses.
+        utxoAll :: UTxO
+        utxoAll = changeUTxO txSet All
+
+        -- The UTxO set that contains only even-parity output addresses.
+        utxoEven :: UTxO
+        utxoEven = changeUTxO txSet Even
+
+        -- The UTxO set that contains only odd-parity output addresses.
+        utxoOdd :: UTxO
+        utxoOdd  = changeUTxO txSet Odd
+
+        txSet :: Set Tx
+        txSet = Set.fromList $ uncurry makeTx <$> (zip txIds txOutputs)
+
+        txIds :: [Hash "Tx"]
+        txIds = makeTxId <$> [minBound .. maxBound]
+
+    -- Creates a transaction from a transaction id and a set of outputs, by
+    -- adding dummy data for fields that are not used by 'changeUTxO'.
+    --
+    -- Ideally, we'd leave all of the unused fields undefined (to assert that
+    -- they should not be evaluated or processed in any way), but since the
+    -- fields of the 'Tx' type are strict, our next best option is to provide
+    -- a minimal value for each field.
+    --
+    makeTx :: Hash "Tx" -> TxOutputs -> Tx
+    makeTx txId TxOutputs {unTxOutputs} = Tx
+        { txId
+        , outputs = unTxOutputs
+        -- Unused fields:
+        , fee = Nothing
+        , resolvedCollateral = []
+        , resolvedInputs = []
+        , metadata = Nothing
+        , withdrawals = Map.empty
+        }
+
+    -- Creates a dummy transaction identifier from a single word.
+    makeTxId :: Word8 -> Hash "Tx"
+    makeTxId i = Hash $ BS.pack [i]
+
+-- | Represents all the outputs of a transaction (both change and non-change).
+--
+newtype TxOutputs = TxOutputs
```

Comment:
I don't get why we go to the trouble of making a `newtype`, but don't add an `Arbitrary` instance, and then use `shrinkAll`, when a simple `[TxOut]` parameter on the property would do the same thing.

Reply:
@rvl Fair enough!

I've added an `Arbitrary` instance for `TxOutputs` in 761ecc21a740e95ef5367a9a41919edfbbf7763a.